pkg/cloud/services/awsnode/cni.go

```diff
@@ -0,0 +1,134 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package awsnode
+
+import (
+	"context"
+	"fmt"
+
+	amazoncni "github.com/aws/amazon-vpc-cni-k8s/pkg/apis/crd/v1alpha1"
+	"sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/awserrors"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	infrav1 "sigs.k8s.io/cluster-api-provider-aws/api/v1alpha3"
+)
+
+func (s *Service) ReconcileCNI() error {
+	if s.scope.SecondaryCidrBlock() == nil {
+		return nil
+	}
+	s.scope.Info("Reconciling awsnode DaemonSet in cluster", "cluster-name", s.scope.Name(), "cluster-namespace", s.scope.Namespace())
+
+	var ds appsv1.DaemonSet
+	if err := s.client.Get(context.Background(), types.NamespacedName{Namespace: "kube-system", Name: "aws-node"}, &ds); err != nil {
+		if !errors.IsNotFound(err) {
+			return err
+		}
+		return ErrCNIMissing
+	}
+
+	sgs, err := s.getSecurityGroups()
+	if err != nil {
+		return err
+	}
+
+	s.scope.Info("for each subnet", "cluster-name", s.scope.Name(), "cluster-namespace", s.scope.Namespace())
+	for _, subnet := range s.secondarySubnets() {
+		var eniConfig amazoncni.ENIConfig
+		if err := s.client.Get(context.Background(), types.NamespacedName{Namespace: v1.NamespaceSystem, Name: subnet.AvailabilityZone}, &eniConfig); err != nil {
+			if !errors.IsNotFound(err) {
+				return err
+			}
+			s.scope.Info("Creating ENIConfig", "cluster-name", s.scope.Name(), "cluster-namespace", s.scope.Namespace(), "subnet", subnet.ID, "availability-zone", subnet.AvailabilityZone)
+			eniConfig = amazoncni.ENIConfig{
+				ObjectMeta: v1.ObjectMeta{
+					Namespace: v1.NamespaceSystem,
+					Name:      subnet.AvailabilityZone,
+				},
+				Spec: amazoncni.ENIConfigSpec{
+					Subnet:         subnet.ID,
+					SecurityGroups: sgs,
+				},
+			}
+
+			if err := s.client.Create(context.Background(), &eniConfig, &client.CreateOptions{}); err != nil {
+				return err
+			}
+		}
+
+		s.scope.Info("Updating ENIConfig", "cluster-name", s.scope.Name(), "cluster-namespace", s.scope.Namespace(), "subnet", subnet.ID, "availability-zone", subnet.AvailabilityZone)
```

Comment:
If we're supporting updates, do the existing but no longer needed `ENIConfig`s need to be deleted?

Reply:
That's a good point. It's also made me think we should be labeling these as being managed by cluster-api so we don't make changes to any created by the user outside of CAPA.

Reply:
Updated 🙂